go/types/value_store.go

```diff
@@ -229,35 +242,115 @@ func (lvs *ValueStore) WriteValue(v Value) Ref {
+	lvs.pendingMu.Lock()
+	defer lvs.pendingMu.Unlock()
+	h := c.Hash()
+	d.Chk.NotZero(height)
+	lvs.pendingPuts[h] = pendingChunk{c, height, hints}
+	lvs.pendingPutSize += uint64(len(c.Data()))
+
+	putChildren := func(parent hash.Hash) (dataPut int) {
+		pc, present := lvs.pendingPuts[parent]
+		d.Chk.True(present)
+		v := DecodeValue(pc.c, lvs)
+		v.WalkRefs(func(grandchildRef Ref) {
+			if pc, present := lvs.pendingPuts[grandchildRef.TargetHash()]; present {
 				lvs.bs.SchedulePut(pc.c, pc.height, pc.hints)
-				delete(lvs.pendingPuts, reachable.TargetHash())
+				dataPut += len(pc.c.Data())
+				delete(lvs.pendingPuts, grandchildRef.TargetHash())
 			}
 		})
-	}()
+		return
+	}
 
-	lvs.setHintsForReachable(v, v.Hash(), true)
-	lvs.set(h, (*presentChunk)(v.Type()), false)
-	lvs.valueCache.Drop(h)
-	return r
+	if height > 1 {
+		v.WalkRefs(func(childRef Ref) {
+			childHash := childRef.TargetHash()
+			if _, present := lvs.pendingPuts[childHash]; present {
+				lvs.pendingParents[h] = height
+			} else {
+				// Shouldn't be able to be in pendingParents without being in pendingPuts
+				_, present := lvs.pendingParents[childHash]
+				d.Chk.False(present)
+			}
+
+			if _, present := lvs.pendingParents[childHash]; present {
+				lvs.pendingPutSize -= uint64(putChildren(childHash))
+				delete(lvs.pendingParents, childHash)
+			}
+		})
+	}
+
+	for lvs.pendingPutSize > lvs.pendingPutMax {
+		var tallest hash.Hash
+		var height uint64 = 0
+		for parent, ht := range lvs.pendingParents {
+			if ht > height {
+				tallest = parent
+				height = ht
+			}
+		}
+		if height == 0 { // This can happen if there are no pending parents
+			var pc pendingChunk
+			for tallest, pc = range lvs.pendingPuts {
+				// Any pendingPut is as good as another in this case, so take the first one
+				break
+			}
+			lvs.bs.SchedulePut(pc.c, pc.height, pc.hints)
+			lvs.pendingPutSize -= uint64(len(pc.c.Data()))
+			delete(lvs.pendingPuts, tallest)
+			continue
+		}
+
+		lvs.pendingPutSize -= uint64(putChildren(tallest))
+		delete(lvs.pendingParents, tallest)
+	}
 }
 
-func (lvs *ValueStore) Flush() {
+func (lvs *ValueStore) Flush(root hash.Hash) {
 	func() {
 		lvs.pendingMu.Lock()
 		defer lvs.pendingMu.Unlock()
-		for _, pc := range lvs.pendingPuts {
+
+		pc, present := lvs.pendingPuts[root]
+		if !present {
+			return
+		}
+
+		put := func(h hash.Hash, pc pendingChunk) uint64 {
 			lvs.bs.SchedulePut(pc.c, pc.height, pc.hints)
+			delete(lvs.pendingPuts, h)
+			return uint64(len(pc.c.Data()))
+		}
+		v := DecodeValue(pc.c, lvs)
+		v.WalkRefs(func(reachable Ref) {
+			if pc, present := lvs.pendingPuts[reachable.TargetHash()]; present {
+				lvs.pendingPutSize -= put(reachable.TargetHash(), pc)
+			}
+		})
+		delete(lvs.pendingParents, root) // If not present, this is idempotent
+		lvs.pendingPutSize -= put(root, pc)
+
+		// Merge in pending hints
+		lvs.cacheMu.Lock()
+		defer lvs.cacheMu.Unlock()
+		for h, entry := range lvs.pendingHints {
```

Comment:
this doesn't look right. it seems like the thing we should be doing here is flushing all hints *except* those for chunks which will remaining in `pendingPuts` after this call.

i would have expected this to be

```
if _, present := lvs.pendingPuts[h]; !present {
  lvs,hintCache[h] = entry
  delete(lvs.pendingHints, h)
} else {
  // no else.
}
```